metagraph/src/graph/annotated_dbg.cpp

```diff
@@ -70,12 +70,50 @@ ::annotate_sequence(std::string_view sequence,
     annotator_->add_labels(indices, labels);
 }
 
+void AnnotatedSequenceGraph
+::annotate_sequences(const std::vector<std::pair<std::string, std::vector<Label>>> &data) {
+    assert(check_compatibility());
+
+    std::vector<std::vector<row_index>> ids(data.size());
+    size_t last = 0;
+    for (size_t t = 0; t < data.size(); ++t) {
+        // if the labels are the same, write indexes to the same array
+        auto &indices = data[t].second == data[last].second ? ids[last] : ids[t];
+        indices.reserve(data[t].first.size());
+
+        graph_->map_to_nodes(data[t].first, [&](node_index i) {
+            if (i > 0)
+                indices.push_back(graph_to_anno_index(i));
+        });
+    }
+
+    std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
Since the thread safety of annotators really depends on how they're implemented, how about we give the dynamic annotators their own mutexes and move this lock into the `add_labels` calls?

Reply:
Currently, it only makes sense for add_kmer_counts to compute ranks in parallel.
For add_labels this doesn't make a difference. May be implemented at some point later.